### PR TITLE
AUTO-37 pipeline for pushing docker images

### PR DESCRIPTION
**Pull request by vilkg**
This PR enables [parallel test execution on CI server](https://jira.dhis2.org/browse/AUTO-17):
- avoiding port collisions by running tests in a docker container. This way port doesn´t have to be expose to host machine and tests can communicate with backend in isolated network;

Adds Jenkinsfile with the pipeline for publishing docker images. For testing I used my docker hub account, but as soon as everything works I will use dhis2 account - it's only a matter of configuring "$DOCKER_HUB_OWNER" variable in jenkins and changing credentials in credentials manager.

 - Multibranch pipeline requires "jenkins-as-code" pipeline definition, hence the Jenkinsfile.

### CODE REVIEW

**Comment by vilkg** [comment]
Thanks for comments @amcgee! 

> Is there a reason you're using docker-compose to build the core image rather than just docker? I prefer running docker build . -t <image_name>:<tag> && docker-compose up over running docker-compose up --build since there's nothing in the docker-compose configuration which would change how the image should be built.

Actually, no, same as yours - personal preference. Are there any differences or added benefits of building with docker build? 

> I think when we're publishing to the dhis2 namespace we probably won't want the image name to also be prefixed with dhis2, so we'd probably want to push the built image to dhis2/core rather than dhis2/dhis2-core.

Very good point, I will change it. 

> There's an implicit dependency between the docker-file and the jenkins script since it assumes the created image is called dhis2-core, I think it would be better to pass the whole image name with tag and maybe have an intelligent default in a .env file.

Good point as well, this should go away when I fix second point

**Comment by vilkg** [comment]
@amcgee I fixed 2 and 3. It this what you had in mind? Had to introduce more variables in Jenkinfile, let me know if you have suggestions on naming, seems a bit complicated to me.

**Comment by amcgee** [comment]
> Are there any differences or added benefits of building with docker build?

For me I think the difference is:
1. That `docker-compose` adds an unnecessary layer of abstraction on top of the build.  It literally just passes through to a `docker build` under the hood, but it's opaque from the call `docker-compose up --build`, since that could be setting a bunch of environment variable or build args or whatever else but you don't know what's actually being done unless you look in the e2e test docker-compose file.  By contrast, if we explicitly build the image we can see in-line what's being done (i.e. just `docker build . -t <sometag>`).  We also no longer need to rely on docker-compose tagging the image.
2. There are two unrelated steps we're conflating into one using the `docker-compose` convenience method of `--build`.  I think it makes sense in the second instance where we're building and running the api testing image, but in the first instance we're separately (a) building an image which we will later do several things to, including pushing to docker hub and (b) spinning up an instance of that image to run tests.

Functionally they're identical, so I'm not going to be ornery about moving to a two-command process, but I do think that they shouldn't be conflated since they're logically distinct steps.

Just my 2c, everything else looks good!

**Comment by amcgee** [comment]
In fact, the `docker compose up -d` in the `build` step could just be directly replaced by the `docker build` command, I believe, since the container will already be started in the following `up` command.  And it would have the added benefit of not leaving a straggling daemon container for Jenkins to (hopefully) clean up.

File: Jenkinsfile-publish-image [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                        sh "IMAGE_NAME=${DOCKER_IMAGE_FULL_NAME} docker-compose -p ${COMPOSE_PARAMETER} up -d --build"
+                    }   
+                }
+            }
+        }
+
+        stage('Run api tests') {
+            steps {
+                dir("dhis-2/dhis-e2e-test") {
+                    sh "TAG=${DOCKER_IMAGE_TAG} docker-compose -p ${COMPOSE_PARAMETER} -f docker-compose.e2e.yml up --build --exit-code-from e2e-test"
```
Comment (amcgee) [inline Jenkinsfile-publish-image]: I'm confused, what is `TAG` here?  It's aliased to `DOCKER_IMAGE_TAG` but used for the tag of a different base image?

Comment (varl) [reply Jenkinsfile-publish-image]: It's used as a short-hand in the `docker-compose.e2e.yml` file: https://github.com/dhis2/dhis2-core/pull/3309/files#diff-b865f401f8b5994344025e530a5e5b37R6

Comment (varl) [reply Jenkinsfile-publish-image]: Same as the `IMAGE_NAME` here: https://github.com/dhis2/dhis2-core/pull/3309/files#diff-d56f0e3b867ad4410c433d69c0da0435R13

**Comment by varl** [comment]
I kept scanning for a `docker build` line somewhere, and it took me a search to find the `--build` at the end of two lines to figure out where the build happens. I'd rather they be on separate lines too as a matter of personal preference. I probably won't delve into these files too often so whatever floats the maintainers raft is OK to me.

**Comment by vilkg** [comment]
I changed from docker-compose --build to docker build. @varl comment convinced me that it's more straightforward with docker build. 

> Is this possible to backport to ... far back?

Yes, that's the plan. I think tests are only in since 2.32, but I will try to back-port it into all supported versions.

**Comment by vilkg** [comment]
@varl @amcgee 

I messed up somewhere with merge, so I created new PR... https://github.com/dhis2/dhis2-core/pull/3326/files. 

Also back-ported to 2.32: https://github.com/dhis2/dhis2-core/pull/3327

**Pull request closed without merging**